invalid pointer

[INFO] [launch]: All log files can be found below /home/vis/.ros/log/2023-08-15-21-27-35-910267-vis-Z590-D-95045
[INFO] [launch]: Default logging verbosity is set to INFO
/home/vis/Desktop/cam/install/usb_cam/share/usb_cam/config/params_1.yaml
[INFO] [usb_cam_node_exe-1]: process started with pid [95046]
[INFO] [show_image.py-2]: process started with pid [95048]
[usb_cam_node_exe-1] [INFO] [1692120456.064841763] [usb_cam]: camera_name value: test_camera
[usb_cam_node_exe-1] [WARN] [1692120456.064921117] [usb_cam]: framerate: 30.000000
[usb_cam_node_exe-1] [INFO] [1692120456.065972218] [usb_cam]: camera calibration URL: package://usb_cam/config/camera_info.yaml
[usb_cam_node_exe-1] [INFO] [1692120456.066341856] [usb_cam]: Starting 'test_camera' (/dev/video0) at 1920x1080 via mmap (mjpeg2rgb) at 30 FPS
[usb_cam_node_exe-1] [INFO] [1692120456.182394506] [usb_cam]: This devices supproted formats:
[usb_cam_node_exe-1] [INFO] [1692120456.182518988] [usb_cam]: 	Motion-JPEG: 1920 x 1080 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182526842] [usb_cam]: 	Motion-JPEG: 1920 x 1080 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182530764] [usb_cam]: 	Motion-JPEG: 1280 x 720 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182534342] [usb_cam]: 	Motion-JPEG: 960 x 540 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182537549] [usb_cam]: 	Motion-JPEG: 640 x 360 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182540552] [usb_cam]: 	Motion-JPEG: 640 x 480 (54 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182543538] [usb_cam]: 	Motion-JPEG: 800 x 600 (54 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182546908] [usb_cam]: 	Motion-JPEG: 1024 x 768 (54 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182549867] [usb_cam]: 	Motion-JPEG: 1280 x 960 (54 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182552914] [usb_cam]: 	Motion-JPEG: 1600 x 1200 (54 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182586956] [usb_cam]: 	Motion-JPEG: 1920 x 1080 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182590304] [usb_cam]: 	Motion-JPEG: 1920 x 1080 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182593376] [usb_cam]: 	YUYV 4:2:2: 1920 x 1080 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182596436] [usb_cam]: 	YUYV 4:2:2: 1920 x 1080 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182599559] [usb_cam]: 	YUYV 4:2:2: 1280 x 720 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182602748] [usb_cam]: 	YUYV 4:2:2: 960 x 540 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182605950] [usb_cam]: 	YUYV 4:2:2: 640 x 360 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182609058] [usb_cam]: 	YUYV 4:2:2: 640 x 480 (54 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182612126] [usb_cam]: 	YUYV 4:2:2: 800 x 600 (54 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182615183] [usb_cam]: 	YUYV 4:2:2: 1024 x 768 (54 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182618240] [usb_cam]: 	YUYV 4:2:2: 1280 x 960 (54 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182621258] [usb_cam]: 	YUYV 4:2:2: 1600 x 1200 (54 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182624468] [usb_cam]: 	YUYV 4:2:2: 1920 x 1080 (60 Hz)
[usb_cam_node_exe-1] [INFO] [1692120456.182627711] [usb_cam]: 	YUYV 4:2:2: 1920 x 1080 (60 Hz)
[usb_cam_node_exe-1] unknown control 'white_balance_temperature_auto'
[usb_cam_node_exe-1] 
[usb_cam_node_exe-1] [INFO] [1692120456.184186976] [usb_cam]: Setting 'white_balance_temperature_auto' to 1
[usb_cam_node_exe-1] [INFO] [1692120456.184205567] [usb_cam]: Setting 'exposure_auto' to 3
[usb_cam_node_exe-1] unknown control 'exposure_auto'
[usb_cam_node_exe-1] 
[usb_cam_node_exe-1] [INFO] [1692120456.185539165] [usb_cam]: Setting 'focus_auto' to 0
[usb_cam_node_exe-1] unknown control 'focus_auto'
[usb_cam_node_exe-1] 
[usb_cam_node_exe-1] [INFO] [1692120456.188846173] [usb_cam]: Timer triggering every 33 ms
[usb_cam_node_exe-1] free(): invalid pointer
[ERROR] [usb_cam_node_exe-1]: process has died [pid 95046, exit code -6, cmd '/home/vis/Desktop/cam/install/usb_cam/lib/usb_cam/usb_cam_node_exe --ros-args -r __node:=usb_cam --params-file /home/vis/Desktop/cam/install/usb_cam/share/usb_cam/config/params_1.yaml'].


I compiled this command 
ros2 launch usb_cam demo_launch.py

--- a/README.md
+++ b/README.md
@@ -77,7 +77,7 @@ ros2 run usb_cam usb_cam_node_exe --ros-args --params-file /path/to/colcon_ws/sr
 
 # launch the usb_cam executable that loads parameters from the same `usb_cam/config/params.yaml` file as above
 # along with an additional image viewer node
-ros2 launch usb_cam demo_launch.py
+ros2 launch usb_cam camera.launch.py
 ```
 ## Launching Multiple usb_cam's
 

--- a/include/usb_cam/formats/mjpeg.hpp
+++ b/include/usb_cam/formats/mjpeg.hpp
@@ -187,11 +187,11 @@ public:
     // Pass src MJPEG image to decoder
     m_result = avcodec_send_packet(m_avcodec_context, m_avpacket);
 
-    av_packet_unref(m_avpacket);
     // If result is not 0, report what went wrong
     if (m_result != 0) {
       std::cerr << "Failed to send AVPacket to decode: ";
       print_av_error_string(m_result);
+      return;
     }
 
     m_result = avcodec_receive_frame(m_avcodec_context, m_avframe_device);
@@ -201,6 +201,7 @@ public:
     } else if (m_result < 0) {
       std::cerr << "Failed to recieve decoded frame from codec: ";
       print_av_error_string(m_result);
+      return;
     }
 
     sws_scale(

--- a/include/usb_cam/formats/mjpeg.hpp
+++ b/include/usb_cam/formats/mjpeg.hpp
@@ -145,6 +145,15 @@ public:
 
   ~MJPEG2RGB()
   {
+    if (m_averror_str) {
+      free(m_averror_str);
+    }
+    if (m_avoptions) {
+      free(m_avoptions);
+    }
+    if (m_avpacket) {
+      free(m_avpacket);
+    }
     if (m_avcodec_context) {
       avcodec_close(m_avcodec_context);
       avcodec_free_context(&m_avcodec_context);
@@ -154,10 +163,6 @@ public:
     }
     if (m_avframe_rgb) {
       av_frame_free(&m_avframe_rgb);
-    }
-    if (m_avpacket) {
-      av_packet_unref(m_avpacket);
-      av_packet_free(&m_avpacket);
     }
     if (m_avparser) {
       av_parser_close(m_avparser);

--- a/include/usb_cam/formats/mjpeg.hpp
+++ b/include/usb_cam/formats/mjpeg.hpp
@@ -75,7 +75,6 @@ public:
     m_avframe_device(av_frame_alloc()),
     m_avframe_rgb(av_frame_alloc()),
     m_avoptions(NULL),
-    m_avpacket(av_packet_alloc()),
     m_averror_str(reinterpret_cast<char *>(malloc(AV_ERROR_MAX_STRING_SIZE)))
   {
     if (!m_avcodec) {
@@ -84,9 +83,6 @@ public:
 
     if (!m_avparser) {
       throw std::runtime_error("Could not find MJPEG parser");
-    }
-    if (!m_avpacket) {
-      throw std::runtime_error("Could not allocate AVPacket");
     }
 
     m_avcodec_context = avcodec_alloc_context3(m_avcodec);
@@ -151,9 +147,6 @@ public:
     if (m_avoptions) {
       free(m_avoptions);
     }
-    if (m_avpacket) {
-      free(m_avpacket);
-    }
     if (m_avcodec_context) {
       avcodec_close(m_avcodec_context);
       avcodec_free_context(&m_avcodec_context);
@@ -179,18 +172,14 @@ public:
     // clear the picture
     memset(dest, 0, m_avframe_device_size);
 
-    #if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(58, 133, 100)
-    // deprecated: https://github.com/FFmpeg/FFmpeg/commit/f7db77bd8785d1715d3e7ed7e69bd1cc991f2d07
-    av_init_packet(m_avpacket);
-    #endif
-
-    av_packet_from_data(
-      m_avpacket,
-      const_cast<uint8_t *>(reinterpret_cast<const uint8_t *>(src)),
-      bytes_used);
+    auto avpacket = av_packet_alloc();
+    av_new_packet(avpacket, bytes_used);
+    memcpy(avpacket->data, src, bytes_used);
 
     // Pass src MJPEG image to decoder
-    m_result = avcodec_send_packet(m_avcodec_context, m_avpacket);
+    m_result = avcodec_send_packet(m_avcodec_context, avpacket);
+
+    av_packet_free(&avpacket);
 
     // If result is not 0, report what went wrong
     if (m_result != 0) {
@@ -234,7 +223,6 @@ private:
   AVFrame * m_avframe_device;
   AVFrame * m_avframe_rgb;
   AVDictionary * m_avoptions;
-  AVPacket * m_avpacket;
   SwsContext * m_sws_context;
   size_t m_avframe_device_size;
   size_t m_avframe_rgb_size;

--- a/src/usb_cam_node.cpp
+++ b/src/usb_cam_node.cpp
@@ -362,13 +362,10 @@ void UsbCamNode::update()
 {
   if (m_camera->is_capturing()) {
     // If the camera exposure longer higher than the framerate period
-    // then that caps the framerate.
-    // auto t0 = now();
+    // then that caps the framerate
     if (!take_and_send_image()) {
       RCLCPP_WARN_ONCE(this->get_logger(), "USB camera did not respond in time.");
     }
-    // auto diff = now() - t0;
-    // INFO(diff.nanoseconds() / 1e6 << " " << int(t0.nanoseconds() / 1e9));
   }
 }
 }  // namespace usb_cam

